test/unit/test-controller-promise.js

```diff
@@ -0,0 +1,308 @@
+      await p.then(thenSpy, thenCatchSpy);
+      await p.then(null, onlyCatchSpy);
+
+      expect(thenSpy).to.not.have.been.called;
+      expect(thenCatchSpy).to.have.been.calledOnce;
+      expect(thenCatchSpy).to.have.been.calledWith(errorObject);
+      expect(onlyCatchSpy).to.have.been.calledOnce;
+      expect(onlyCatchSpy).to.have.been.calledWith(errorObject);
+    });
+
+    it('should accept rejected calls to `finally`', async () => {
+      const catchSpy = sandbox.spy();
+      const finallySpy = sandbox.spy();
+      const failureObject = {};
+      const rejectedControllerPromise = new ControllerPromise(
+        Promise.reject(failureObject)
+      );
+
+      // The catch in this line prevents Promise.reject from breaking the test
+      await rejectedControllerPromise.catch(catchSpy).finally(finallySpy);
+
+      expect(catchSpy).to.have.been.calledOnce;
+      expect(finallySpy).to.have.been.calledOnce;
+    });
+
+    it('should accept resolved calls to `finally`', async () => {
+      const finallySpy = sandbox.spy();
+      const thenSpy = sandbox.spy();
+      const successObject = {};
+      const resolvedControllerPromise = new ControllerPromise(
+        Promise.resolve(successObject)
+      );
+
+      await resolvedControllerPromise.then(thenSpy).finally(finallySpy);
+
+      expect(thenSpy).to.have.been.calledWith(successObject);
+      expect(finallySpy).to.have.been.calledOnce;
+    });
+
+    it('should pass errors beyond `finally` to `catch` blocks', async () => {
+      const finallySpy = sandbox.spy();
+      const catchSpy = sandbox.spy();
+      const failureObject = {};
+      const rejectedControllerPromise = new ControllerPromise(
+        Promise.reject(failureObject)
+      );
+
+      await rejectedControllerPromise.finally(finallySpy).catch(catchSpy);
+
+      expect(finallySpy).to.have.been.calledOnce;
+      expect(catchSpy).to.have.been.calledWith(failureObject);
+    });
+
+    it('should accept long then chains', async () => {
+      const p = new ControllerPromise(Promise.resolve(1));
+      const two = p.then(x => x + 1);
+
+      const three = two.then(x => x + 1).then(x => 'hello world ' + x);
+
+      expect(await three).to.equal('hello world 3');
+    });
+  });
+
+  describe('retryable behavior', () => {
+    it('should allow promises to be retryable', async () => {
+      const valueFunction = sandbox.stub();
+      valueFunction.resolves(6);
+      valueFunction
+        .onCall(0)
+        .resolves(0)
+        .onCall(1)
+        .resolves(1)
+        .onCall(2)
+        .resolves(2)
+        .onCall(3)
+        .resolves(3)
+        .onCall(4)
+        .resolves(4)
+        .onCall(5)
+        .resolves(5);
+
+      const p = new ControllerPromise(
+        Promise.resolve(0),
+        getWaitFunction(valueFunction)
+      );
+
+      expect(await p).to.equal(0);
+      expect(await p.waitForValue(x => x == 1)).to.equal(1);
```

Comment:
This feels trivial for testing `waitForValue()`. Can we use a timeout or sleep?

Reply:
Ah sorry the sequence of `waitForValue` calls kind of makes this unclear. The calls to `waitForValue` don't call `valueFunction` directly, so these check that the value eventually equals each of expected values. The `setInterval` in the `getWaitFunction` helper calls `valueFunction` every 4ms and advances it, like the sleep you're suggesting. I'll remove the series of values and only expect the last one to make that clear.

Also I found that this didn't behave like I intended if there are multiple calls to `waitForValue` since each interval would advance the `valueFunction`, so I changed `getWaitFunction` so each call to `waitForValue` has its own `valueFunction`. Thanks!

Reply:
Ah neat! :) Thanks